lua/ttt2/extensions/net.lua

```diff
@@ -0,0 +1,99 @@
+---
+-- net stream extension
+-- @author saibotk
+
+AddCSLuaFile()
+
+-- Stream network message name constant
+local NETMSG_STREAM = "TTT2_NET_STREAM"
+
+-- Size to split the network stream at (currently a bit lower than the max value, just to have some buffer)
+net.STREAM_FRAGMENTATION_SIZE = 65400
+
+-- Stream cache variables
+net.stream_cache = {}
+net.stream_callbacks = {}
+
+-- Add the network string for streaming data
+if SERVER then
+	util.AddNetworkString(NETMSG_STREAM)
+end
+
+---
+-- Initiates a stream message, usually for data that can be longer than
+-- the 64kb limit of a single net message. This will split up the data and send them in
+-- smaller fragments. The data will be converted (with sPON) to an encoded string during this process.
+--
+-- @param string messageId a unique message id similar to the network strings
+-- @param table data the data table to send, this will be reconstructed at the client.
+-- @param table|player|nil client SERVERSIDE only! Optional, use it to send a stream to a single client or a group of clients.
+-- @realm shared
+function net.SendStream(messageId, data, client)
+	local encodedString = pon.encode(data)
+	local split = string.SplitAtSize(encodedString, net.STREAM_FRAGMENTATION_SIZE)
+	for i = 1, #split do
+		net.Start(NETMSG_STREAM)
+		-- Write the messageId
+		net.WriteUInt(util.CRC(messageId), 32)
+		-- Write if there are still fragments coming after this one
+		net.WriteBool(i < #split)
+		-- Write the actual data fragment as a string, which internally will also send its size
+		net.WriteString(split[i])
+
+		if SERVER then
+			if client then
+				net.Send(client)
+			else
+				net.Broadcast()
+			end
+		else
+			net.SendToServer()
+		end
+	end
+end
+
+---
+-- Receive a stream message, usually for data that can be longer than
+-- the 64kb limit of a single net message. This will register a callback
+-- for a specific messageId and execute it when the stream was received and
+-- the data is reconstructed from all fragments.
+--
+-- @param string messageId a unique message id similar to the network strings
+-- @param function callback This is the function that is called after the data was received.
+-- @realm shared
+function net.ReceiveStream(messageId, callback)
+	-- has to be saved as string, otherwise the key lookups will fail on the table
+	local msg = tostring(util.CRC(messageId))
+	net.stream_callbacks[msg] = callback
```

Comment:
CS: missing empty line in front

Reply:
you could inline this as well

Reply:
inlining this would make it harder to read, especially for table keys